Mathlib/Tactic/Simproc/Divisors.lean

```diff
@@ -0,0 +1,66 @@
+/-
+Copyright (c) 2025 Paul Lezeau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Lezeau, Bhavik Mehta
+-/
+import Mathlib.NumberTheory.Divisors
+import Mathlib.Util.Qq
+import Mathlib.Data.Finset.Sort
+
+/-! # Divisor Simprocs
+
+This file implements simprocs to compute various objects related to divisors:
+- `Nat.divisorsEq`: computes `Nat.divisors n` for explicit values of `n`
+- `Nat.properDivisorsEq`: computes `Nat.properDivisors n` for explicit values of `n`
+
+-/
+
+--TODO: these simprocs can probably be made a lot more efficient. See the discussion in #23026.
+
+open Lean Meta Qq
+
+/-- The `Nat.divisorsEq` computes the finset `Nat.divisors n` when `n` is a natural number
+literal. -/
+simproc_decl Nat.divisorsEq (Nat.divisors _) := fun e => do
+  unless e.isAppOfArity `Nat.divisors 1 do return .continue
+  let some n ← fromExpr? e.appArg! | return .continue
+  let rhsListQ : List Q(ℕ) := (n.divisors.sort (· ≤ ·)).map fun n => (Lean.toExpr n : Q(ℕ))
+  let rhs := mkSetLiteralQ q(Finset ℕ) rhsListQ
+  let pf ← Meta.mkDecideProof (← mkEq e rhs)
+  return .done {expr := rhs, proof? := pf }
+
+/-- The `Nat.properDivisorsEq ` computes the finset `Nat.properDivisorsEq  n` when `n` is a natural
+number literal. -/
+simproc_decl Nat.properDivisorsEq (Nat.properDivisors _) := fun e => do
+  unless e.isAppOfArity `Nat.properDivisors 1 do return .continue
+  let some n ← fromExpr? e.appArg! | return .continue
+  let rhsListQ : List Q(ℕ) := (n.properDivisors.sort (· ≤ ·)).map fun n => (Lean.toExpr n : Q(ℕ))
+  let rhs := mkSetLiteralQ q(Finset ℕ) rhsListQ
+  let pf ← Meta.mkDecideProof (← mkEq e rhs)
+  return .done {expr := rhs, proof? := pf }
+
+example : Nat.divisors 1710 = {1, 2, 3, 5, 6, 9, 10, 15, 18, 19, 30, 38, 45, 57,
```

Comment:
How high can you go before you get in trouble? Maybe add a test with a big factorial and a bit power.

Reply:
Seems like `7!` already causes some issues! In terms of powers, `2^20` runs relatively fast.

Reply:
fwiw, asymptotically, primorials have the most divisors, so those might be what you want to use to stress test

Reply:
Added a demo with primorials